lib/security/JweToken.ts

```diff
@@ -14,6 +15,49 @@ type EncryptDelegate = (data: Buffer, jwk: PublicKey) => Promise<Buffer>;
  * This class hides the JOSE and crypto library dependencies to allow support for additional crypto algorithms.
  */
 export default class JweToken extends JoseToken {
+
+  // used for verification if a JSON Serialized JWS was given
+  private readonly protected: string | undefined;
+  private unprotected: {[key: string]: any} | undefined;
+  private readonly encrypted_key: string | undefined;
+  private readonly iv: string | undefined;
+  private readonly tag: string | undefined;
+  private readonly aad: string | undefined;
+  private readonly isFlattenedJSONSerialized: boolean;
+
+  public constructor (content: string | object, protected cryptoFactory: CryptoFactory) {
+    super(content, cryptoFactory);
+    const jsonContent: any = content;
+    this.isFlattenedJSONSerialized = false;
+    if (typeof jsonContent === 'object' &&
+    'ciphertext' in jsonContent && typeof jsonContent.ciphertext === 'string' &&
+    'iv' in jsonContent && typeof jsonContent.iv === 'string' &&
+    'tag' in jsonContent && typeof jsonContent.tag === 'string' &&
+    (('protected' in jsonContent && typeof jsonContent.protected === 'string') ||
+    ('unprotected' in jsonContent && typeof jsonContent.unprotected === 'object'))) {
```

Comment:
Sanity: this will incorrectly pass the conditional check if `protected in JsonContent` and `jsonContent.unprotected` is a `string`

Reply:
If I understand correctly, you're asking if it should pass when one parameter is the wrong type, while the other is correct.

Per JWE's [rfc 7.2.1](https://tools.ietf.org/html/rfc7516#section-7.2.1) "Additional members can be present in both the JSON objects defined above; if not understood by implementations encountering them, they MUST be ignored."

Its a bit ambiguous to me if an understood member like "protected" of the incorrect type is an incorrect JSON serialized JWE, or just an additional member with a coliding name. Either way I need to fix the behavior, so I suppose it would be more defensive to assume its entirely incorrect and reject the JWE?